File: tools/token_bench/programs/04_concurrency/concurrency.py
from concurrent.futures import ThreadPoolExecutor
import queue
import threading

def fib(n):
    if n < 2:
        return n
    return fib(n - 1) + fib(n - 2)

with ThreadPoolExecutor() as pool:
    futures = [pool.submit(fib, 27) for _ in range(4)]
    r = sum(f.result() for f in futures)
    print(f"parallel: {r}")

with ThreadPoolExecutor() as pool:
    results = list(pool.map(fib, [27, 27, 27, 27]))
    total = sum(results)
    print(f"pmap: {total}")

q = queue.Queue(10)

def producer():
    for i in range(10):
        q.put(i * i)

t = threading.Thread(target=producer)
t.start()
for _ in range(10):
    print(q.get())
t.join()

File: tools/token_bench/programs/04_concurrency/concurrency.cs
using System;
using System.Collections.Concurrent;
using System.Linq;
using System.Threading.Tasks;

class Program {
    static int Fib(int n) {
        if (n < 2) return n;
        return Fib(n - 1) + Fib(n - 2);
    }

    static void Main() {
        var tasks = Enumerable.Range(0, 4).Select(_ => Task.Run(() => Fib(27))).ToArray();
        Task.WaitAll(tasks);
        int r = tasks.Sum(t => t.Result);
        Console.WriteLine($"parallel: {r}");

        var inputs = new[] {27, 27, 27, 27};
        var results = inputs.AsParallel().Select(Fib).ToArray();
        int total = results.Sum();
        Console.WriteLine($"pmap: {total}");

        var queue = new BlockingCollection<int>(10);
        Task.Run(() => {
            for (int i = 0; i < 10; i++) queue.Add(i * i);
            queue.CompleteAdding();
        });
        foreach (var v in queue.GetConsumingEnumerable()) {
            Console.WriteLine(v);
        }
    }
}
